server/src/main/java/org/elasticsearch/index/mapper/TextFieldMapper.java

```diff
@@ -680,7 +681,10 @@ public IntervalsSource intervals(String text, int maxGaps, boolean ordered,
         @Override
         public Query phraseQuery(TokenStream stream, int slop, boolean enablePosIncrements) throws IOException {
             String field = name();
-            if (indexPhrases && slop == 0 && hasGaps(stream) == false) {
+            // we can't use the index_phrases shortcut with slop, if there are gaps in the stream,
+            // or if the incoming token stream is the output of a token graph due to
+            // https://issues.apache.org/jira/browse/LUCENE-8916
+            if (indexPhrases && slop == 0 && hasGaps(stream) == false && stream.hasAttribute(BytesTermAttribute.class) == false) {
```

Comment:
Do we also want to add `stream.hasAttribute(BytesTermAttribute.class)` condition for `multiPhraseQuery` below?

Reply:
I don't know much about graph token filters, but I wonder if another way to check that index_phrases is not used with synonyms is to modify `hasGaps` to check also that `posIncAtt.getPositionIncrement() != 0`

Reply:
This is a workaround for an implementation detail, namely that we represent the paths through a token graph using `BytesTermAttribute` instead of `CharTermAttribute` - once LUCENE-8916 has been fixed then this workaround should stop working, so I think we probably should keep checking for the implementation.

Reply:
We don't need this for multiPhraseQuery, because the results of calling `getFiniteStrings()` on a token graph are always simple token streams with no stacked tokens.  The bug appears because in `MatchQuery.MatchQueryBuilder.analyzeGraphBoolean()` we iterate through all possible paths in the graph and send them back through `createFieldQuery()`, which because the paths are all simple will end up delegating to either `FieldType#termQuery()` or `FieldType#phraseQuery` - `multiPhraseQuery` will never be encountered in this second loop, so there's no need to check for the attribute.